### PR TITLE
Polymer 2.0 Support

### LINKED ISSUE #156: Polymer 2.0 Support
- Use latest `v2.0.0-rcX`
- Aim for Hybrid if feasible
- Add "variants" for "2.x" dependencies in `bower.json`
- Plan is to release this as `2.0` and then deprecate the Polymer v1 support.
  - There is going to be a breaking API change because of we need start using named slots instead of class selectors for child elements.

- Follow steps from https://www.polymer-project.org/2.0/docs/upgrade for more details.

### PR DESCRIPTION
**Pull request by Saulis**
Fixes #156

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-upload/166)

### CODE REVIEW

**Comment by manolo** [comment]
Reviewed 2 of 6 files at r1, 2 of 4 files at r2.
Review status: 4 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[HYBRID.md, line 1 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KgyGmLAMgthn5XJ4gXM:-KgyGmLAMgthn5XJ4gXN:bj2d1fk) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ff6d6f238d9530076e5b6cea1a0415363cb314fc/HYBRID.md#L1)):*
> ```Markdown
> # Transforming this element to hybrid element.
> ```

I guess we dont need file in the repo

---

*[vaadin-upload.html, line 102 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KgyHjbwxmgVSlW8cCVN:-KgyHjbwxmgVSlW8cCVO:b-kzrpc1) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ff6d6f238d9530076e5b6cea1a0415363cb314fc/vaadin-upload.html#L102)):*
> ```HTML
>       :host([dragover-valid]) #dropLabel {
>         color: var(--primary-color, #00B4F0);
>         @apply(--vaadin-upload-drop-label-dragover);
> ```

you need to remove parentheses,  replace `@apply(...)` by `@apply ...`

---

*[demo/advanced.html, line 42 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KgyI2LEHTm05z-8rvAj:-KgyI2LEHTm05z-8rvAk:b-eh5z7n) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ff6d6f238d9530076e5b6cea1a0415363cb314fc/demo/advanced.html#L42)):*
> ```HTML
>         <template is="dom-bind">
>           <vaadin-upload id="fileEventsDemo" files="{{files}}">
>             <div slot=file-list>
> ```

please add double quotes, also in other demos your removed them

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

**Comment by manolo** [comment]
@juniperbelmont seems that you have not signed the CLA, pls follow the instructions and do it.

---

Reviewed 1 of 4 files at r2.
Review status: 5 of 9 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-:-KgyIODUKziZaAdlfWjM:b-tqy31p)*

**Comment by Saulis** [comment]
Just a comment: review will continue once we continue working on the Polymer 2.0 support.

File: bower.json [hunk 1]
```diff
@@ -27,24 +27,46 @@
     "package.json",
     "wct.conf.js"
   ],
+  "dependencies": {
+    "polymer": "^1.0.0",
```
Comment (stramel) [inline bower.json]: This will definitely have to be at least 1.7.0 since you are using slots here.

**Comment by manolo** [comment]
I just rebased with master and checked that dependencies installs, etc.

---

Review status: 2 of 8 files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[bower.json, line 31 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KgxuT-lzdQyY0PuMCd4-r5-31:-KhY3aGPEo3KfJKARfiQ:bb6un1j) ([raw file](https://github.com/vaadin/vaadin-upload/blob/c835fa53a326bb1a1643357c24ca492214b3ef8b/bower.json#L31)):*
<details><summary><i>Previously, stramel (Michael Stramel) wrote…</i></summary><blockquote>

This will definitely have to be at least 1.7.0 since you are using slots here.
</blockquote></details>

Done

---

*[vaadin-upload.html, line 155 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KhR-j5rQQ0AVJfRbs5b:-KhR-j5rQQ0AVJfRbs5c:b84nwxe) ([raw file](https://github.com/vaadin/vaadin-upload/blob/9747900362e7c0149c93c479f96ee908b9bb3766/vaadin-upload.html#L155)):*
<details><summary><i alt="summary">Quoted 20 lines of code…</i></summary>
>         <div id="dropLabel" hidden$="[[nodrop]]">
>           <slot name="drop-label-icon">
>             <iron-icon id="dropLabelIcon" icon="vaadin-upload:file-upload"></iron-icon>
>           </slot>
>           <slot name="drop-label">
>             [[_i18nPlural(maxFiles, i18n.dropFiles, i18n.dropFiles.*)]]
>           </slot>
>         </div>
>       </div>
>     </div>
>     <slot name="file-list">
>       <div id="fileList">
>         <template is="dom-repeat" items="[[files]]" as="file">
>           <vaadin-upload-file file="[[file]]"></vaadin-upload-file>
>         </template>
>       </div>
>     </slot>
>     <slot></slot>
>     <input type="file" id="fileInput" on-change="_onFileInputChange" hidden accept$="{{accept}}" multiple$="[[_isMultiple(maxFiles)]]">
>     <paper-ripple id="dragRipple" noink></paper-ripple>
</details>
>   </template>
>  
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>       <template>
>         <template is="dom-bind">
>           <vaadin-upload id="fileEventsDemo" files="{{files}}">
>             <div slot="file-list"></details>


---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-:-KhY3e3Iek3PQd3p0AGZ:bz1oqo2)*

**Comment by manolo** [comment]
Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[vaadin-upload.html, line 155 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KhR-j5rQQ0AVJfRbs5b:-KhY3ytlkMb2IoooCQfd:b-kscniq) ([raw file](https://github.com/vaadin/vaadin-upload/blob/9747900362e7c0149c93c479f96ee908b9bb3766/vaadin-upload.html#L155)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

>         <div id="dropLabel" hidden$="[[nodrop]]">
>           <slot name="drop-label-icon">
>             <iron-icon id="dropLabelIcon" icon="vaadin-upload:file-upload"></iron-icon>
>           </slot>
>           <slot name="drop-label">
>             [[_i18nPlural(maxFiles, i18n.dropFiles, i18n.dropFiles.*)]]
>           </slot>
>         </div>
>       </div>
>     </div>
>     <slot name="file-list">
>       <div id="fileList">
>         <template is="dom-repeat" items="[[files]]" as="file">
>           <vaadin-upload-file file="[[file]]"></vaadin-upload-file>
>         </template>
>       </div>
>     </slot>
>     <slot></slot>
>     <input type="file" id="fileInput" on-change="_onFileInputChange" hidden accept$="{{accept}}" multiple$="[[_isMultiple(maxFiles)]]">
>     <paper-ripple id="dragRipple" noink></paper-ripple>
>   </template>
>  
>       <template>
>         <template is="dom-bind">
>           <vaadin-upload id="fileEventsDemo" files="{{files}}">
>             <div slot="file-list">

</blockquote></details>

No idea why this comment is here, pls ignore it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

File: .travis.yml [hunk 2]
```diff
@@ -13,12 +13,17 @@ addons:
...
+  - polymer install --variants
 
 before_script:
   - gulp lint
-  - ([ "$TRAVIS_EVENT_TYPE" = "pull_request" ] || TRAVIS_BRANCH=quick/${TRAVIS_BUILD_ID} xvfb-run -s '-screen 0 1024x768x24' wct)
+  - ([ "$TRAVIS_EVENT_TYPE" = "pull_request" ] || TRAVIS_BRANCH=quick/${TRAVIS_BUILD_ID} xvfb-run -s '-screen 0 1024x768x24' polymer test)
 
 script:
+  # test against Polymer 1
+  - xvfb-run -s '-screen 0 1024x768x24' wct
```
Comment (stramel) [inline .travis.yml]: Use `polymer test` here and remove the moving of `bower_components`

`polymer test` will test both variants.

File: bower.json [hunk 3]
```diff
@@ -28,23 +28,45 @@
...
     "mock-http-request": "philikon/MockHttpRequest#master",
-    "paper-toast": "PolymerElements/paper-toast#^1.0.0",
-    "test-fixture": "PolymerElements/test-fixture#^1.0.0",
-    "web-component-tester": "Polymer/web-component-tester#^5.0.0"
+    "paper-toast": "^1.0.0",
+    "web-component-tester": "^6.0.0-prerelease.7",
+    "webcomponentsjs": "^0.7.0"
+  },
+  "variants": {
+    "polymer2": {
```
Comment (stramel) [inline bower.json]: It looks like your variants are backward. Should be `1.x` variant and mainline should be 2.0. After an official version of Polymer 2.0 is tagged, using `^1.9||^2.0` (this was the syntax discussed as the way to move forward post 2.0 tag of polymer)

https://www.polymer-project.org/2.0/docs/devguide/hybrid-elements#dependency-variants

**Comment by stramel** [comment]
Would now be a good time to split out vaadin-upload into single and multiple uploads as discussed in #149 ? I could submit a PR to the 2.0 branch for that.

File: bower.json (hunk 3 above)
Comment (Saulis) [inline bower.json]: Our plan is to release the hybrid versions of our elements with minimal breaking changes to the previous release (including dependencies). Afterwards, we'll do a new release which drops the hybrid support and aims purely Polymer `2.0`

File: .travis.yml (hunk 2 above)
Comment (Saulis) [inline .travis.yml]: Good idea!

**Comment by Saulis** [comment]
@stramel sure, if you've got time to do that. I encourage chatting with the team in our Gitter chat first to make sure all the nitty-gritty details about the designs align.

File: test/index.html [hunk 4]
```diff
@@ -10,14 +10,14 @@
...
-    WCT._config.environmentScripts.push('webcomponentsjs/webcomponents.js');
+    WCT._config.environmentScripts.push('webcomponentsjs/webcomponents-lite.js');
 
     WCT.loadSuites([
       'upload.html',
       'adding-files.html',
       'file-list.html'
     ].reduce(function(suites, suite) {
-      return suites.concat([suite, suite + '?dom=shadow']);
+      return suites.concat([suite, suite + '?wc-shadydom=true&dom=shadow']);
```
Comment (stramel) [inline test/index.html]: This line doesn't make sense to me. Enabling shady Dom and using Shadow Dom? 

It has been recommended to me to use 

test.html?wc-shadydom=true&wc-ce=true
test.html?dom=shadow

File: .travis.yml [hunk 5]
```diff
@@ -13,12 +13,11 @@ addons:
 
 install:
   - npm install
-  - npm install -g bower
-  - bower install
+  - polymer install --variants
 
 before_script:
   - gulp lint
```
Comment (stramel) [inline .travis.yml]: Can I suggest adding `polymer lint` in here as well

File: bower.json [hunk 6]
```diff
@@ -28,23 +28,45 @@
...
   "dependencies": {
-    "iron-icon": "PolymerElements/iron-icon#^1.0.0",
-    "iron-iconset-svg": "polymerelements/iron-iconset-svg#^1.0.0",
-    "paper-button": "PolymerElements/paper-button#^1.0.0",
-    "paper-icon-button": "PolymerElements/paper-icon-button#^1.0.0",
-    "paper-progress": "PolymerElements/paper-progress#^1.0.0",
-    "paper-ripple": "PolymerElements/paper-ripple#~1.0.0",
-    "paper-styles": "PolymerElements/paper-styles#^1.0.0",
-    "polymer": "Polymer/polymer#^1.7.0"
+    "iron-icon": "^1.0.0",
```
Comment (stramel) [inline bower.json]: Is there a reason you guys switched away from specifying the GitHub repo like how the Polymer team does?

Comment (limonte) [inline bower.json]: - easier to read
- DRY

File: .travis.yml (hunk 5 above)
Comment (limonte) [inline .travis.yml]: Added, thx!

Comment (stramel) [inline .travis.yml]: May want to use polymer-2-hybrid rules. Thanks for adding the linting :+1:

Comment (limonte) [inline .travis.yml]: Done :+1:

File: test/index.html (hunk 4 above)
Comment (Saulis) [inline test/index.html]: Yeah, that might be a confusing, but the idea is to use `dom=shadow` when running tests in Polymer v1 to enable shadow and `wc-shadydom=true` to enable shady when running tests in Polymer v2.

**Comment by Saulis** [comment]
:lgtm: 

---

Reviewed 1 of 6 files at r1, 1 of 4 files at r2, 1 of 3 files at r3, 2 of 4 files at r6, 1 of 1 files at r8, 1 of 3 files at r9, 2 of 3 files at r10, 1 of 1 files at r11, 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[vaadin-upload.html, line 102 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KgyHjbwxmgVSlW8cCVN:-KiYe0XKT5SZrHIJlG2I:b-896fix) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ff6d6f238d9530076e5b6cea1a0415363cb314fc/vaadin-upload.html#L102)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

you need to remove parentheses,  replace `@apply(...)` by `@apply ...`
</blockquote></details>

Done.

---

*[vaadin-upload.html, line 155 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KhR-j5rQQ0AVJfRbs5b:-KiYe5Xfegd_cDsETA6X:b-896fix) ([raw file](https://github.com/vaadin/vaadin-upload/blob/9747900362e7c0149c93c479f96ee908b9bb3766/vaadin-upload.html#L155)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

No idea why this comment is here, pls ignore it.
</blockquote></details>

Done.

---

*[demo/advanced.html, line 42 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KgyI2LEHTm05z-8rvAj:-KiYe2K2lStm1WXKVwh-:b-896fix) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ff6d6f238d9530076e5b6cea1a0415363cb314fc/demo/advanced.html#L42)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

please add double quotes, also in other demos your removed them
</blockquote></details>

Done.

---

*[HYBRID.md, line 1 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KgyGmLAMgthn5XJ4gXM:-KiYe3QkcbMBjMQFvyL4:b-896fix) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ff6d6f238d9530076e5b6cea1a0415363cb314fc/HYBRID.md#L1)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I guess we dont need file in the repo
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-:-KiYeFBV6u6xAILu9C3L:bg7igkv)*

**Comment by tomivirkki** [comment]
*[README.md, line 85 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KiYo4JrVxDoc_I1Eejd:-KiYo4JrVxDoc_I1Eeje:bq9j0js) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ddfe28d67edb2f0104991e3373b1948a2d411aa8/README.md#L85)):*
> ```Markdown
>   - Polymer1-compatible version:
> 
>     - API documentation: http://127.0.0.1:8000/components/vaadin-upload/
> ```

I guess the ports aren't fixed. For me, the mainline is 8081 and polymer2 is 8000.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

**Comment by tomivirkki** [comment]
The "Custom File List" demo (under "Advanced Usage") doesn't show in P2 version (getting some errors in the console on that page). The navigation bar also looks unstyled in P2 version demos.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-:-KiYqjJTkt1duDtqO1No:baqn3fi)*

**Comment by limonte** [comment]
> The "Custom File List" demo (under "Advanced Usage") doesn't show in P2 version (getting some errors in the console on that page). The navigation bar also looks unstyled in P2 version demos.

Fixed :+1:

---

Review status: 10 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[README.md, line 85 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KiYo4JrVxDoc_I1Eejd:-KiYsXs-_7oLu5xamqzP:by8doh) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ddfe28d67edb2f0104991e3373b1948a2d411aa8/README.md#L85)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

I guess the ports aren't fixed. For me, the mainline is 8081 and polymer2 is 8000.
</blockquote></details>

That's true, `8000` and `8001` aren't fixed, those are default. Should I add some note about that to the end of this section in README? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-:-KiZ6xSpRAFkJtDhTbk8:bm4y6jq)*

**Comment by limonte** [comment]
Review status: 9 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[test/index.html, line 20 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KiTzgHTPXqNkVZlmRJx-r10-20:-KiZ7jS93x2swcuI83D2:ba7a5dn) ([raw file](https://github.com/vaadin/vaadin-upload/blob/f4c2f0cc6a0385075caa50d5293f6a212b92ae4f/test/index.html#L20)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Yeah, that might be a confusing, but the idea is to use `dom=shadow` when running tests in Polymer v1 to enable shadow and `wc-shadydom=true` to enable shady when running tests in Polymer v2.
</blockquote></details>

I added clarifying comment to make this clear in future

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

**Comment by tomivirkki** [comment]
*[README.md, line 85 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KiYo4JrVxDoc_I1Eejd:-KiZBCKSKerow0XSHwES:bhn6vix) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ddfe28d67edb2f0104991e3373b1948a2d411aa8/README.md#L85)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

That's true, `8000` and `8001` aren't fixed, those are default. Should I add some note about that to the end of this section in README? 
</blockquote></details>

Well, I don't know why they default to different ports on my environment. Could be a bit confusing to someone who tries this and ends up on a blank page.
The readme could also use mention prior to this step that you need to run `polymer install --variants`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

**Comment by limonte** [comment]
Review status: 8 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[README.md, line 85 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KiYo4JrVxDoc_I1Eejd:-KiZKJeiL_rRAlOo8IMa:b-qhft13) ([raw file](https://github.com/vaadin/vaadin-upload/blob/ddfe28d67edb2f0104991e3373b1948a2d411aa8/README.md#L85)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Well, I don't know why they default to different ports on my environment. Could be a bit confusing to someone who tries this and ends up on a blank page.
The readme could also use mention prior to this step that you need to run `polymer install --variants`.
</blockquote></details>

added `polymer install --variants` and the note about ports

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

**Comment by tomivirkki** [comment]
*[demo/advanced.html, line 34 at r13](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KiZLr1n4eBdbAKcjOhJ:-KiZLr1n4eBdbAKcjOhK:bl4a1lv) ([raw file](https://github.com/vaadin/vaadin-upload/blob/dccaa9df7e7212f06d3c07d28304075736cda0d1/demo/advanced.html#L34)):*
> ```HTML
>     <h3>Custom File List</h3>
>     <demo-snippet>
>       <dom-bind>
> ```

The wrapping `<template>` tag is still needed by the `<demo-snippet>`. Without it, the sources are not visible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

**Comment by Saulis** [comment]
Review status: 8 of 14 files reviewed at latest revision, 10 unresolved discussions.

---

*[demo/advanced.html, line 34 at r13](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-KiZLr1n4eBdbAKcjOhJ:-KiZQfDPA02UVa_9dyJe:b-896fix) ([raw file](https://github.com/vaadin/vaadin-upload/blob/dccaa9df7e7212f06d3c07d28304075736cda0d1/demo/advanced.html#L34)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

The wrapping `<template>` tag is still needed by the `<demo-snippet>`. Without it, the sources are not visible.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166)*

**Comment by tomivirkki** [comment]
:lgtm:

---

Reviewed 1 of 6 files at r1, 1 of 4 files at r6, 1 of 1 files at r8, 2 of 3 files at r10, 1 of 1 files at r11, 2 of 3 files at r12, 5 of 6 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-:-KiZSy4mJ3zS7rUK-_VB:bnfp4nl)*

**Comment by limonte** [comment]
Thanks for nice catches, Tomi!

:lgtm: 

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-upload/166#-:-KiZW0MUhAKTGsKJ-guA:b-scjc6r)*

**Pull request merged**